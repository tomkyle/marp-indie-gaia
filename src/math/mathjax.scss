/**
 * Disable thickening strokes in print media (e.g. PDF).
 *
 * https://github.com/mathjax/MathJax-src/blob/2dd53ce6c8af3c9cceba0baf014ea9b065130774/ts/output/svg/Wrappers/TextNode.ts#L49-L53
 * https://github.com/marp-team/marp-core/issues/287
 */
@media print {
  mjx-container[jax='SVG'] {
    use[data-c],
    path[data-c] {
      stroke-width: 0;
    }
  }
}
