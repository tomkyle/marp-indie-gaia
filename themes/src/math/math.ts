import marpitPlugin from '@marp-team/marpit/plugin'
import { Marp } from '../marp'
import { getMathContext, setMathContext } from './context'
import * as katex from './katex'
import * as mathjax from './mathjax'

export type MathPreferredLibrary = 'mathjax' | 'katex'

export interface MathOptionsInterface {
  lib?: MathPreferredLibrary
  katexOption?: Record<string, unknown>
  katexFontPath?: string | false
}

export type MathOptions = boolean | MathPreferredLibrary | MathOptionsInterface

const defaultLibrary = 'mathjax' as const
const getLibrary = (opts: MathOptionsInterface) => opts.lib ?? defaultLibrary

export const markdown = marpitPlugin((md) => {
  const marp: Marp = md.marpit
  const opts: MathOptions | undefined = marp.options.math

  if (!opts) return

  const parsedOpts =
    typeof opts !== 'object'
      ? { lib: typeof opts === 'string' ? opts : undefined }
      : opts

  // Define `math` global directive to choose preferred library
  Object.defineProperty(marp.customDirectives.global, 'math', {
    value: (math: unknown): { math?: MathPreferredLibrary } => {
      if (math === 'katex' || math === 'mathjax') return { math }
      return {}
    },
  })

  // Initialize
  const { parse, parseInline } = md

  const initializeMathContext = () => {
    if (getMathContext(marp).processing) return false

    setMathContext(marp, () => ({
      enabled: false,
      options: parsedOpts,
      processing: true,
      katexMacroContext: {
        ...((parsedOpts.katexOption?.macros as any) || {}),
      },
      mathjaxContext: null,
    }))

    return true
  }

  const parseWithMath = <F extends (this: any, ...args: any[]) => any>(
    func: F,
  ) => {
    return function (this: ThisType<F>, ...args: Parameters<F>) {
      const initialized = initializeMathContext()

      try {
        return func.apply(this, args)
      } finally {
        if (initialized) {
          setMathContext(marp, (ctx) => ({ ...ctx, processing: false }))
        }
      }
    }
  }

  md.parse = parseWithMath(parse)
  md.parseInline = parseWithMath(parseInline)

  const enableMath = () =>
    setMathContext(marp, (ctx) => ({ ...ctx, enabled: true }))

  // Inline
  md.inline.ruler.after('escape', 'marp_math_inline', (state, silent) => {
    const ret = parseInlineMath(state, silent)
    if (ret) enableMath()

    return ret
  })

  // Block
  md.block.ruler.after(
    'blockquote',
    'marp_math_block',
    (state, start, end, silent) => {
      const ret = parseMathBlock(state, start, end, silent)
      if (ret) enableMath()

      return ret
    },
    { alt: ['paragraph', 'reference', 'blockquote', 'list'] },
  )

  // Renderer
  md.core.ruler.after(
    'marpit_directives_global_parse',
    'marp_math_directive',
    () => {
      const { enabled } = getMathContext(marp)
      if (!enabled) return

      const preffered: MathPreferredLibrary | undefined = (marp as any)
        .lastGlobalDirectives.math

      setMathContext(marp, (ctx) => ({
        ...ctx,
        options: {
          ...ctx.options,
          lib: preffered ?? parsedOpts.lib ?? defaultLibrary,
        },
      }))
    },
  )

  const getPreferredLibrary = () => {
    const { options } = getMathContext(marp)
    return getLibrary(options) === 'mathjax' ? mathjax : katex
  }

  const getRenderer = (type: 'inline' | 'block') => (tokens: any, idx: any) =>
    getPreferredLibrary()[type](marp)(tokens, idx)

  md.renderer.rules.marp_math_inline = getRenderer('inline')
  md.renderer.rules.marp_math_block = getRenderer('block')
})

export const css = (marpit: any): string | null => {
  const { enabled, options } = getMathContext(marpit)
  if (!enabled) return null

  switch (getLibrary(options)) {
    case 'mathjax':
      return mathjax.css(marpit)
    case 'katex':
      return katex.css(options.katexFontPath)
  }
}

// ---

function isValidDelim(state, pos = state.pos) {
  const ret = { openable: true, closable: true }
  const { posMax, src } = state
  const prev = pos > 0 ? src.charCodeAt(pos - 1) : -1
  const next = pos + 1 <= posMax ? src.charCodeAt(pos + 1) : -1

  if (next === 0x20 || next === 0x09) ret.openable = false
  if (prev === 0x20 || prev === 0x09 || (next >= 0x30 && next <= 0x39)) {
    ret.closable = false
  }

  return ret
}

function parseInlineMath(state, silent) {
  const { src, pos } = state
  if (src[pos] !== '$') return false

  const addPending = (stt: string) => (state.pending += stt)
  const found = (manipulation: () => void, newPos: number) => {
    if (!silent) manipulation()
    state.pos = newPos
    return true
  }

  const start = pos + 1
  if (!isValidDelim(state).openable) return found(() => addPending('$'), start)

  let match = start
  while ((match = src.indexOf('$', match)) !== -1) {
    let dollarPos = match - 1
    while (src[dollarPos] === '\\') dollarPos -= 1

    if ((match - dollarPos) % 2 === 1) break
    match += 1
  }

  if (match === -1) return found(() => addPending('$'), start)
  if (match - start === 0) return found(() => addPending('$$'), start + 1)
  if (!isValidDelim(state, match).closable) {
    return found(() => addPending('$'), start)
  }

  return found(() => {
    const token = state.push('marp_math_inline', 'math', 0)
    token.markup = '$'
    token.content = src.slice(start, match)
  }, match + 1)
}

function parseMathBlock(state, start, end, silent) {
  const { blkIndent, bMarks, eMarks, src, tShift } = state
  let pos = bMarks[start] + tShift[start]
  let max = eMarks[start]

  if (pos + 2 > max || src.slice(pos, pos + 2) !== '$$') return false
  if (silent) return true

  pos += 2

  let firstLine = src.slice(pos, max)
  let lastLine
  let found = firstLine.trim().slice(-2) === '$$'

  if (found) firstLine = firstLine.trim().slice(0, -2)

  let next = start
  for (; !found; ) {
    next += 1
    if (next >= end) break

    pos = bMarks[next] + tShift[next]
    max = eMarks[next]
    if (pos < max && tShift[next] < blkIndent) break

    const target = src.slice(pos, max).trim()

    if (target.slice(-2) === '$$') {
      found = true
      lastLine = src.slice(pos, src.slice(0, max).lastIndexOf('$$'))
    }
  }

  state.line = next + 1

  const token = state.push('marp_math_block', 'math', 0)
  token.block = true
  token.content = ''
  token.map = [start, state.line]
  token.markup = '$$'

  if (firstLine?.trim()) token.content += `${firstLine}\n`
  token.content += state.getLines(start + 1, next, tShift[start], true)
  if (lastLine?.trim()) token.content += lastLine

  return true
}
