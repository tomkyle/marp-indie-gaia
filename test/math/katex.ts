/**
 * All test cases are ported from markdown-it-katex.
 *
 * @see https://github.com/waylonflinn/markdown-it-katex/blob/master/test/fixtures/default.txt
 */

import MarkdownIt from 'markdown-it'
import { markdown as mathPlugin } from '../../src/math/math'

const countMath = (stt) => stt.split('class="katex"').length - 1
const countBlockMath = (stt) => stt.split('class="katex-display"').length - 1

describe('markdown-it math plugin for KaTeX', () => {
  const md = new MarkdownIt()

  md.marpit = { options: { math: 'katex' } }

  // Mock Marpit instance
  md.use(() => {
    Object.assign(md.marpit, {
      customDirectives: { global: {} },
      lastGlobalDirectives: {},
    })
    md.core.ruler.push('marpit_directives_global_parse', () => {
      // no ops
    })
  })
  md.use(mathPlugin)

  it('renders simple inline math', () => {
    const rendered = md.render('$1+1 = 2$')
    expect(countMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('renders simple block math', () => {
    const rendered = md.render('$$1+1 = 2$$')
    expect(countBlockMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('renders math without whitespace before and after delimiter', () => {
    const rendered = md.render('foo$1+1 = 2$bar')
    expect(countMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('renders math even when it starts with a negative sign', () => {
    const rendered = md.render('foo$-1+1 = 0$bar')
    expect(countMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('does not render with empty content', () => {
    const rendered = md.render('aaa $$ bbb')
    expect(countMath(rendered)).toBe(0)
    expect(rendered).toMatchSnapshot()
  })

  it('requires a closing delimiter to render math', () => {
    const rendered = md.render('aaa $5.99 bbb')
    expect(countMath(rendered)).toBe(0)
    expect(rendered).toMatchSnapshot()
  })

  it('does not allow paragraph break in inline math', () => {
    const rendered = md.render('foo $1+1\n\n= 2$ bar')
    expect(countMath(rendered)).toBe(0)
    expect(rendered).toMatchSnapshot()
  })

  it('does not process apparent markup in inline math', () => {
    const rendered = md.render('foo $1 *i* 1$ bar')
    expect(countMath(rendered)).toBe(1)
    expect(rendered).not.toContain('<em>')
    expect(rendered).toMatchSnapshot()
  })

  it('renders block math with indented up to 3 spaces', () => {
    const rendered = md.render('   $$\n   1+1 = 2\n   $$')
    expect(countBlockMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('does not render block math with indented up to 4 spaces (code block)', () => {
    const rendered = md.render('    $$\n    1+1 = 2\n    $$')
    expect(countBlockMath(rendered)).toBe(0)
    expect(rendered).toContain('<code>')
    expect(rendered).toMatchSnapshot()
  })

  it('renders multiline inline math', () => {
    const rendered = md.render('foo $1 + 1\n= 2$ bar')
    expect(countMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('renders multiline display math', () => {
    const rendered = md.render('$$\n\n  1\n+ 1\n\n= 2\n\n$$')
    expect(countBlockMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('allows to place text immediately after inline math', () => {
    const rendered = md.render('$n$-th order')
    expect(countMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('renders block math with self-closing at the end of document', () => {
    const rendered = md.render('$$\n1+1 = 2')
    expect(countBlockMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('can appear both maths in lists', () => {
    const rendered = md.render('* $1+1 = 2$\n* $$\n  1+1 = 2\n  $$')
    expect(countMath(rendered)).toBe(2)
    expect(countBlockMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('renders block math written in one line', () => {
    const rendered = md.render('$$1+1 = 2$$')
    expect(countBlockMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('renders block math composed multiple lines with starting/ending expression on delimited lines', () => {
    const rendered = md.render('$$[\n[1, 2]\n[3, 4]\n]$$')
    expect(countBlockMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('does not render math when delimiters are escaped', () => {
    const rendered = md.render('Foo \\$1$ bar\n\\$\\$\n1\n\\$\\$')
    expect(countMath(rendered)).toBe(0)
    expect(rendered).toMatchSnapshot()
  })

  it('does not render math when numbers are followed closing inline math', () => {
    const rendered = md.render(
      "Thus, $20,000 and USD$30,000 won't parse as math.",
    )
    expect(countMath(rendered)).toBe(0)
    expect(rendered).toMatchSnapshot()
  })

  it('requires non whitespace to right of opening inline math', () => {
    const rendered = md.render(
      'For some Europeans, it is 2$ for a can of soda, not 1$.',
    )
    expect(countMath(rendered)).toBe(0)
    expect(rendered).toMatchSnapshot()
  })

  it('requires non whitespace to left of closing inline math', () => {
    const rendered = md.render(
      'I will give you $20 today, if you give me more $ tomorrow.',
    )
    expect(countMath(rendered)).toBe(0)
    expect(rendered).toMatchSnapshot()
  })

  it('does not recognize inline block math', () => {
    const rendered = md.render(
      "It's well know that $$1 + 1 = 3$$ for sufficiently large 1.",
    )
    expect(countMath(rendered)).toBe(0)
    expect(rendered).toMatchSnapshot()
  })

  it('recognizes escaped delimiters in math mode', () => {
    const rendered = md.render('Money adds: $\\$X + \\$Y = \\$Z$.')
    expect(countMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })

  it('recognizes multiline escaped delimiters in math module', () => {
    const rendered = md.render(
      'Weird-o: $\\displaystyle{\\begin{pmatrix} \\$ & 1\\\\\\$ \\end{pmatrix}}$.',
    )
    expect(countMath(rendered)).toBe(1)
    expect(rendered).toMatchSnapshot()
  })
})
