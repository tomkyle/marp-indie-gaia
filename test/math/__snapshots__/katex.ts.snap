// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown-it math plugin for KaTeX allows to place text immediately after inline math 1`] = `
"<p><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi></mrow><annotation encoding="application/x-tex">n</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.4306em;"></span><span class="mord mathnormal">n</span></span></span></span>-th order</p>
"
`;

exports[`markdown-it math plugin for KaTeX can appear both maths in lists 1`] = `
"<ul>
<li><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>+</mo><mn>1</mn><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">1+1 = 2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">2</span></span></span></span></li>
<li>
<p><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mn>1</mn><mo>+</mo><mn>1</mn><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">1+1 = 2
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">2</span></span></span></span></span></p></li>
</ul>
"
`;

exports[`markdown-it math plugin for KaTeX does not allow paragraph break in inline math 1`] = `
"<p>foo $1+1</p>
<p>= 2$ bar</p>
"
`;

exports[`markdown-it math plugin for KaTeX does not process apparent markup in inline math 1`] = `
"<p>foo <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>∗</mo><mi>i</mi><mo>∗</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">1 *i* 1</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">∗</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6595em;"></span><span class="mord mathnormal">i</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">∗</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span></span></span></span> bar</p>
"
`;

exports[`markdown-it math plugin for KaTeX does not recognize inline block math 1`] = `
"<p>It's well know that $$1 + 1 = 3$$ for sufficiently large 1.</p>
"
`;

exports[`markdown-it math plugin for KaTeX does not render block math with indented up to 4 spaces (code block) 1`] = `
"<pre><code>$$
1+1 = 2
$$
</code></pre>
"
`;

exports[`markdown-it math plugin for KaTeX does not render math when delimiters are escaped 1`] = `
"<p>Foo $1$ bar
$$
1
$$</p>
"
`;

exports[`markdown-it math plugin for KaTeX does not render math when numbers are followed closing inline math 1`] = `
"<p>Thus, $20,000 and USD$30,000 won't parse as math.</p>
"
`;

exports[`markdown-it math plugin for KaTeX does not render with empty content 1`] = `
"<p>aaa $$ bbb</p>
"
`;

exports[`markdown-it math plugin for KaTeX recognizes escaped delimiters in math mode 1`] = `
"<p>Money adds: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi mathvariant="normal">$</mi><mi>X</mi><mo>+</mo><mi mathvariant="normal">$</mi><mi>Y</mi><mo>=</mo><mi mathvariant="normal">$</mi><mi>Z</mi></mrow><annotation encoding="application/x-tex">\\$X + \\$Y = \\$Z</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8333em;vertical-align:-0.0833em;"></span><span class="mord">$</span><span class="mord mathnormal" style="margin-right:0.07847em;">X</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.8056em;vertical-align:-0.0556em;"></span><span class="mord">$</span><span class="mord mathnormal" style="margin-right:0.22222em;">Y</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.8056em;vertical-align:-0.0556em;"></span><span class="mord">$</span><span class="mord mathnormal" style="margin-right:0.07153em;">Z</span></span></span></span>.</p>
"
`;

exports[`markdown-it math plugin for KaTeX recognizes multiline escaped delimiters in math module 1`] = `
"<p>Weird-o: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mstyle scriptlevel="0" displaystyle="true"><mrow><mo fence="true">(</mo><mtable rowspacing="0.16em" columnalign="center center" columnspacing="1em"><mtr><mtd><mstyle scriptlevel="0" displaystyle="false"><mi mathvariant="normal">$</mi></mstyle></mtd><mtd><mstyle scriptlevel="0" displaystyle="false"><mn>1</mn></mstyle></mtd></mtr><mtr><mtd><mstyle scriptlevel="0" displaystyle="false"><mi mathvariant="normal">$</mi></mstyle></mtd></mtr></mtable><mo fence="true">)</mo></mrow></mstyle></mrow><annotation encoding="application/x-tex">\\displaystyle{\\begin{pmatrix} \\$ &amp; 1\\\\\\$ \\end{pmatrix}}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:2.4em;vertical-align:-0.95em;"></span><span class="mord"><span class="minner"><span class="mopen delimcenter" style="top:0em;"><span class="delimsizing size3">(</span></span><span class="mord"><span class="mtable"><span class="col-align-c"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.45em;"><span style="top:-3.61em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">$</span></span></span><span style="top:-2.41em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">$</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.95em;"><span></span></span></span></span></span><span class="arraycolsep" style="width:0.5em;"></span><span class="arraycolsep" style="width:0.5em;"></span><span class="col-align-c"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:1.45em;"><span style="top:-3.61em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span></span></span></span></span></span><span class="mclose delimcenter" style="top:0em;"><span class="delimsizing size3">)</span></span></span></span></span></span></span>.</p>
"
`;

exports[`markdown-it math plugin for KaTeX renders block math composed multiple lines with starting/ending expression on delimited lines 1`] = `
"<p><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mo stretchy="false">[</mo><mo stretchy="false">[</mo><mn>1</mn><mo separator="true">,</mo><mn>2</mn><mo stretchy="false">]</mo><mo stretchy="false">[</mo><mn>3</mn><mo separator="true">,</mo><mn>4</mn><mo stretchy="false">]</mo><mo stretchy="false">]</mo></mrow><annotation encoding="application/x-tex">[
[1, 2]
[3, 4]
]</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mopen">[[</span><span class="mord">1</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.1667em;"></span><span class="mord">2</span><span class="mclose">]</span><span class="mopen">[</span><span class="mord">3</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.1667em;"></span><span class="mord">4</span><span class="mclose">]]</span></span></span></span></span></p>"
`;

exports[`markdown-it math plugin for KaTeX renders block math with indented up to 3 spaces 1`] = `
"<p><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mn>1</mn><mo>+</mo><mn>1</mn><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">1+1 = 2
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">2</span></span></span></span></span></p>"
`;

exports[`markdown-it math plugin for KaTeX renders block math with self-closing at the end of document 1`] = `"<p><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mn>1</mn><mo>+</mo><mn>1</mn><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">1+1 = 2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">2</span></span></span></span></span></p>"`;

exports[`markdown-it math plugin for KaTeX renders block math written in one line 1`] = `
"<p><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mn>1</mn><mo>+</mo><mn>1</mn><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">1+1 = 2
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">2</span></span></span></span></span></p>"
`;

exports[`markdown-it math plugin for KaTeX renders math even when it starts with a negative sign 1`] = `
"<p>foo<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>−</mo><mn>1</mn><mo>+</mo><mn>1</mn><mo>=</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">-1+1 = 0</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">−</span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">0</span></span></span></span>bar</p>
"
`;

exports[`markdown-it math plugin for KaTeX renders math without whitespace before and after delimiter 1`] = `
"<p>foo<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>+</mo><mn>1</mn><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">1+1 = 2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">2</span></span></span></span>bar</p>
"
`;

exports[`markdown-it math plugin for KaTeX renders multiline display math 1`] = `
"<p><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mn>1</mn><mo>+</mo><mn>1</mn><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">
  1
+ 1

= 2

</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">2</span></span></span></span></span></p>"
`;

exports[`markdown-it math plugin for KaTeX renders multiline inline math 1`] = `
"<p>foo <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>+</mo><mn>1</mn><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">1 + 1
= 2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">2</span></span></span></span> bar</p>
"
`;

exports[`markdown-it math plugin for KaTeX renders simple block math 1`] = `
"<p><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mn>1</mn><mo>+</mo><mn>1</mn><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">1+1 = 2
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">2</span></span></span></span></span></p>"
`;

exports[`markdown-it math plugin for KaTeX renders simple inline math 1`] = `
"<p><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>+</mo><mn>1</mn><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">1+1 = 2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">2</span></span></span></span></p>
"
`;

exports[`markdown-it math plugin for KaTeX requires a closing delimiter to render math 1`] = `
"<p>aaa $5.99 bbb</p>
"
`;

exports[`markdown-it math plugin for KaTeX requires non whitespace to left of closing inline math 1`] = `
"<p>I will give you $20 today, if you give me more $ tomorrow.</p>
"
`;

exports[`markdown-it math plugin for KaTeX requires non whitespace to right of opening inline math 1`] = `
"<p>For some Europeans, it is 2$ for a can of soda, not 1$.</p>
"
`;
